File: packages/audioplayers_android_exo/android/src/main/kotlin/xyz/luan/audioplayers/player/ClickTrackConfig.kt
package xyz.luan.audioplayers.player

/**
 * Metronome click-track configuration, mirrored from the Dart side via the
 * `setClickTrack` method channel call.
 *
 * The beat grid lives in *media time*: `anchorMs + offsetMs + n * beatPeriod(bpm)`.
 * The processor consuming this sits before the Signalsmith time-stretcher, so
 * media time is what its frames are measured in and playback speed never
 * affects click placement.
 *
 * Immutable by design: the config reference is swapped atomically
 * (@Volatile) between the platform thread and the audio playback thread.
 */
data class ClickTrackConfig(
    val enabled: Boolean,
    /** Original BPM of the song (grid tempo), NOT the sped-up value. */
    val bpm: Int,
    /** Song-time ms of a known beat (tap-to-align anchor); null = grid from 0. */
    val anchorMs: Long?,
    /** Additional grid shift in song-time ms (user nudge). */
    val offsetMs: Long,
    val beatsPerBar: Int,
    /** 1 = quarters only, 2 = eighths, 3 = triplets, 4 = sixteenths. */
    val pulsesPerBeat: Int,
    /** Click gain 0.0..1.0 relative to the song. */
    val volume: Float,
)

File: packages/audioplayers_darwin/darwin/audioplayers_darwin/Sources/audioplayers_darwin/ClickTrackConfig.swift
import Foundation

/// Metronome click-track configuration, mirrored from the Dart side via the
/// `setClickTrack` method channel call.
///
/// The beat grid lives in *media time*: `anchorMs + offsetMs + n * beatPeriod(bpm)`.
/// The tap consuming this processes source audio in item time (before AVPlayer's
/// rate/time-pitch stage), so media time is what its frames are measured in and
/// playback speed never affects click placement.
///
/// Immutable by design: the config value is handed across threads as a whole
/// (see `ClickTrackTapContext`) — an exact mirror of the Kotlin
/// `ClickTrackConfig` in audioplayers_android_exo.
struct ClickTrackConfig: Equatable {
  let enabled: Bool
  /// Original BPM of the song (grid tempo), NOT the sped-up value.
  let bpm: Int
  /// Song-time ms of a known beat (tap-to-align anchor); nil = grid from 0.
  let anchorMs: Int64?
  /// Additional grid shift in song-time ms (user nudge).
  let offsetMs: Int64
  let beatsPerBar: Int
  /// 1 = quarters only, 2 = eighths, 3 = triplets, 4 = sixteenths.
  let pulsesPerBeat: Int
  /// Click gain 0.0..1.0 relative to the song.
  let volume: Float
}
